Add .NET cryptography primitives

heaths opened the pull request:
Partly resolves #3545 by adding an RSA implementation.

heaths commented:
/cc @clairernovotny

heaths commented on sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/CryptographyClient.cs:
```diff
@@ -1592,6 +1593,62 @@ public virtual VerifyResult VerifyData(SignatureAlgorithm algorithm, Stream data
+        /// The <see cref="CryptographyClient"/> will attempt to download the public key synchronously.
+        /// </remarks>
+        /// <exception cref="RequestFailedException">The server returned an error. See <see cref="Exception.Message"/> for details returned from the server.</exception>
+        public virtual RSA CreateRSA(CancellationToken cancellationToken = default)
```
@KrzysztofCwalina @tg-msft @schaabs there are separate sync and async calls for the initial fetch attempt of the public key; otherwise, waiting to fetch inside the crypto primitive will always be sync over network.

Also, is returning the abstract base `RSA` here fine? One normally should not interact with the derivatives.

heaths commented:
Replaces #25178

heaths commented on sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/RSAKeyVault.cs:
```diff
@@ -0,0 +1,140 @@
+    /// <summary>
+    /// Provides an implementation of the RSA algorithm backed by Azure Key Vault.
+    /// </summary>
+    internal class RSAKeyVault : RSA
```
Talking with @clairenovotny, it'd be nice to expose the public key as well.

heaths replied:
Talking with @schaabs it makes sense to return this class (make it public) so we don't later paint ourselves in a corner if we choose to add functionality like exposing the public key, which you could get from the certificate now anyway.

heaths committed:
Add recorded tests, implement ECDsa support

azure-sdk commented:
**API change check**

APIView has identified API level changes in this PR and created following API reviews.

[Azure.Security.KeyVault.Keys](https://apiview.dev/Assemblies/Review/6589c1188e774f8eae3f58cb195c3d1b)

heaths committed:
Remove ECDsa support

Not currently needed, and we are discussing some changes to make it a better API before release.

christothes commented on sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/CryptographyClient.cs:
```diff
@@ -1592,6 +1601,64 @@ public virtual VerifyResult VerifyData(SignatureAlgorithm algorithm, Stream data
+
+            try
+            {
+                Initialize(GetOperation, cancellationToken);
```
Why not this, like the other methods?
```suggestion
                Initialize(nameof(CreateRSA), cancellationToken);
```

heaths replied:
Because I want to telemeter the `get` operation. Everything else is strictly client-side. That they called `CreateRSA` is captured just above.

christothes commented on sdk/keyvault/Azure.Security.KeyVault.Keys/src/Cryptography/RSAKeyVault.cs:
```diff
@@ -0,0 +1,189 @@
+        /// <param name="padding">The padding mode.</param>
+        /// <returns>The encrypted data.</returns>
+        /// <exception cref="NotSupportedException">The <paramref name="padding"/> is not supported.</exception>
+        /// <exception cref="RequestFailedException"></exception>
```
empty value

heaths committed:
Resolve PR feedback

heaths committed:
Add CHANGELOG entry

Soren-Knudsen commented:
@christothes: Any news for when [4.6.0-beta.1 (Unreleased)](https://github.com/heaths/azure-sdk-for-net/blob/111e47ed904bcb95dbaba033ad8f46a4dc34fec6/sdk/keyvault/Azure.Security.KeyVault.Keys/CHANGELOG.md#460-beta1-unreleased) will be released?

heaths commented:
Before the end of this month.
